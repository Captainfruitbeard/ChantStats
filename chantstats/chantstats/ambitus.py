from enum import Enum
from music21.interval import Interval

__all__ = ["get_ambitus", "AmbitusType"]


class AmbitusType(str, Enum):
    """
    Possible ambitus types for pieces.
    """

    AUTHENTIC = "authentic"
    PLAGAL = "plagal"
    UNDEFINED = "undefined"


def get_ambitus(*, note_of_final, lowest_note):
    if note_of_final is None:
        return AmbitusType.UNDEFINED

    interval = Interval(note_of_final, lowest_note)
    if 0 >= interval.semitones >= -4:
        return AmbitusType.AUTHENTIC
    elif -5 >= interval.semitones > -12:
        return AmbitusType.PLAGAL
    else:
        raise Exception(  # pragma: no cover
            "Check the logic in the ambitus calculation! "
            "We expect the lowest note to be less than an octave "
            "below the main final."
        )
