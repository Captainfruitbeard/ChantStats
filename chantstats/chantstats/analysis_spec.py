import os
import shutil
from abc import ABCMeta
from enum import Enum

from .dendrogram2 import Dendrogram
from .logging import logger
from .modal_category import ModalCategory, GroupingByModalCategory

__all__ = ["FullAnalysisSpec"]


class EnumWithDescription(str, Enum):
    def __new__(cls, name, desc, **kwargs):
        obj = str.__new__(cls, name)
        obj._value_ = name
        obj._description = desc
        return obj

    @property
    def description(self) -> str:
        """
        Returns
        -------
        str
            Long form description of this enum suitable for use in plot titles etc.
        """
        return self._description


class RepertoireAndGenreType(EnumWithDescription):
    PLAINCHANT_SEQUENCES = ("plainchant_sequences", "Plainchant Sequences", ("chant", "sequences"))
    RESPONSORIAL_CHANTS = ("responsorial_chants", "Responsorial Chants", ("chant", "responsorial_chants"))
    ORGANA = ("organa", "Organa", ("organum", ""))

    def __new__(cls, name, desc, path_stubs, **kwargs):
        obj = str.__new__(cls)
        obj._value_ = name
        obj._description = desc
        obj.path_stub_1 = path_stubs[0]
        obj.path_stub_2 = path_stubs[1]
        return obj


class BaseAnalysisFunc(metaclass=ABCMeta):
    def __call__(self, item):
        raise NotImplementedError()


class AnalysisFuncPCFreqs(BaseAnalysisFunc):
    def __init__(self, abs_or_rel):
        assert abs_or_rel in ["abs_freqs", "rel_freqs"]
        self.abs_or_rel = abs_or_rel

    def __call__(self, item, *, unit):
        if unit == UnitType.PCS:
            return getattr(item.pc_freqs, self.abs_or_rel)
        elif unit == UnitType.MODE_DEGREES:
            return getattr(item.mode_degree_freqs, self.abs_or_rel)
        else:
            raise RuntimeError("Unexpected unit: {self.unit}")


class AnalysisType(EnumWithDescription):
    PC_FREQS = ("pc_freqs", "PC Frequencies")
    TENDENCY = ("tendency", "Tendency")
    APPROACHES_AND_DEPARTURES = ("approaches_and_departures", "Approaches & Departures")
    LEAPS_AND_MELODIC_OUTLINES = ("leaps_and_melodic_outlines", "Leaps & Melodic Outlines")

    @property
    def analysis_func(self):
        if self.value == "pc_freqs":
            return AnalysisFuncPCFreqs("rel_freqs")
        else:
            return BaseAnalysisFunc()


class UnitType(EnumWithDescription):
    PCS = ("pcs", "pitch classes")
    MODE_DEGREES = ("mode_degrees", "mode degrees")


class ModeType(EnumWithDescription):
    BY_FINAL = ("by_final", "by final")
    AUTHENTIC_MODES = ("authentic_modes", "authentic modes")
    PLAGAL_MODES = ("plagal_modes", "plagal modes")

    def get_description(self, *, final):
        # TODO: check that final is a valid pitch class
        if self.value == "by_final":
            return f"{final}-final"
        elif self.value == "authentic_modes":
            return f"{final}-authentic"
        elif self.value == "plagal_modes":
            return f"{final}-plagal"
        else:
            raise RuntimeError(f"Unexpected value: {self.value!r}")

    def get_subfolder(self, *, final):
        # TODO: check that final is a valid pitch class
        if self.value == "by_final":
            return f"{final}_final"
        elif self.value == "authentic_modes":
            return f"{final}_authentic"
        elif self.value == "plagal_modes":
            return f"{final}_plagal"
        else:
            raise RuntimeError(f"Unexpected value: {self.value!r}")


class FullAnalysisSpecOLD:
    def __init__(self, *, repertoire_and_genre, analysis, unit, mode):
        self.repertoire_and_genre = RepertoireAndGenreType(repertoire_and_genre)
        self.analysis = AnalysisType(analysis)
        self.unit = UnitType(unit)
        self.mode = ModeType(mode)

        self.analysis_func = self.analysis.analysis_func

    def get_description(self, *, final):
        # TODO: assert that `final` is a valid pitch class
        return f"{self.analysis.description}, {self.repertoire_and_genre.description}, {self.mode.get_description(final=final)}"

    def get_description_NEW(self, *, modal_category):
        assert isinstance(modal_category, ModalCategory)
        return f"{self.analysis.description}, {self.repertoire_and_genre.description}, {modal_category.descr}"

    def __repr__(self):
        return (
            f"<FullAnalysisSpecOLD: "
            f"repertoire_and_genre={self.repertoire_and_genre.value!r}, "
            f"analysis={self.analysis.value!r}, "
            f"unit={self.unit.value!r}, "
            f"mode={self.mode.value!r}, "
            ">"
        )

    def output_path(self, *, root_dir, final):
        return os.path.join(
            root_dir,
            self.repertoire_and_genre.path_stub_1,
            self.analysis.value,
            self.repertoire_and_genre.path_stub_2,
            self.unit.value,
            self.mode.value,
            self.mode.get_subfolder(final=final),
        )

    def export_results(self, *, output_root_dir, grouping, p_cutoff, overwrite=False, sort_freqs_ascending=False):
        raise DeprecationWarning(
            "This method has been moved to the class GroupingByModalCategory. "
            "You should call grouping.export_results(analysis_spec=..., ...)."
        )


class FullAnalysisSpec:
    def __init__(self, *, repertoire_and_genre, analysis, unit):
        self.repertoire_and_genre = RepertoireAndGenreType(repertoire_and_genre)
        self.analysis = AnalysisType(analysis)
        self.unit = UnitType(unit)
        # self.mode = ModeType(mode)

        self.analysis_func = self.analysis.analysis_func

    def get_description(self, *, modal_category):
        assert isinstance(modal_category, ModalCategory)
        return f"{self.analysis.description}, {self.repertoire_and_genre.description}, {modal_category.descr}"

    def __repr__(self):
        return (
            f"<FullAnalysisSpec: "
            f"repertoire_and_genre={self.repertoire_and_genre.value!r}, "
            f"analysis={self.analysis.value!r}, "
            f"unit={self.unit.value!r}, "
            ">"
        )

    def output_path(self, *, root_dir, modal_category):
        return os.path.join(
            root_dir,
            self.repertoire_and_genre.path_stub_1,
            self.analysis.value,
            self.repertoire_and_genre.path_stub_2,
            self.unit.value,
            modal_category.output_path_stub,
        )

    def export_results(self, *, output_root_dir, grouping, p_cutoff, overwrite=False, sort_freqs_ascending=False):
        raise DeprecationWarning(
            "This method has been moved to the class GroupingByModalCategory. "
            "You should call grouping.export_results(analysis_spec=..., ...)."
        )
