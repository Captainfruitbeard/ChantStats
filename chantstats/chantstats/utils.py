import os
import sh

from .logging import logger

__all__ = ["list_directory_tree"]


def list_directory_tree(root_dir):
    """
    List contents of `root_dir` in a tree-like format.

    Returns
    -------
    str
    """
    if not os.path.exists(root_dir):
        logger.error(f"Cannot list tree under non-existent directory: '{root_dir}'")
        return
    cwd = os.getcwd()
    os.chdir(root_dir)
    res = sh.tree().stdout.decode("utf-8")
    os.chdir(cwd)
    return res
