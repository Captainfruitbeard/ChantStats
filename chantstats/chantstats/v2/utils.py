import numpy as np
from enum import Enum

__all__ = ["EnumWithDescription"]


class EnumWithDescription(str, Enum):
    def __new__(cls, name, desc, **kwargs):
        obj = str.__new__(cls, name)
        obj._value_ = name
        obj._description = desc
        return obj

    @property
    def description(self) -> str:
        """
        Returns
        -------
        str
            Long form description of this enum suitable for use in plot titles etc.
        """
        return self._description


def is_close_to_zero_or_100(x):
    return np.isclose(x, 0.0) or np.isclose(x, 100.0)
