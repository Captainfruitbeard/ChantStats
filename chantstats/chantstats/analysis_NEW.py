import os
import pickle
from collections import defaultdict
from tqdm import tqdm

from .analysis_spec import RepertoireAndGenreType, AnalysisType, AnalysisFuncPCFreqs
from .dendrogram2 import Dendrogram
from .logging import logger
from .modal_category import ModalCategoryType, GroupingByModalCategory
from .unit import UnitType
from .plainchant_sequence_monomodal_sections import extract_monomodal_sections

__all__ = ["prepare_analysis_inputs"]


def recursive_defaultdict():
    return defaultdict(recursive_defaultdict)


def get_recursive_subdict(the_dict, keys):
    first_key = keys[0]
    if len(keys) == 1:
        return the_dict[first_key]
    elif len(keys) > 1:
        return get_recursive_subdict(the_dict[first_key], keys[1:])
    else:
        raise RuntimeError()


def convert_to_nested_dict(d):
    if not isinstance(d, dict):
        return d
    else:
        return {key: convert_to_nested_dict(val) for key, val in d.items()}


class AnalysisResultCollection:
    def __init__(self):
        self._results = defaultdict(dict)
        self._results_nested = recursive_defaultdict()  # this is redudant (could compute on the fly), but nice to have

    def __getitem__(self, path_stubs):
        return self._results[path_stubs]

    def _get_path_stubs(self, repertoire_and_genre, analysis, unit, modal_category):
        rep_and_genre = RepertoireAndGenreType(repertoire_and_genre)
        analysis = AnalysisType(analysis)
        unit = UnitType(unit)
        return (
            rep_and_genre.output_path_stub_1,
            analysis.output_path_stub,
            rep_and_genre.output_path_stub_2,
            unit.output_path_stub,
            modal_category.output_path_stub_1,
            modal_category.output_path_stub_2,
        )

    def insert_results(self, repertoire_and_genre, analysis, unit, modal_category, results_key, value):
        path_stubs = self._get_path_stubs(repertoire_and_genre, analysis, unit, modal_category)
        get_recursive_subdict(self._results_nested, path_stubs)[results_key] = value
        self._results[path_stubs][results_key] = value

    def to_dict(self):
        return dict(self._results)  # convert from defaultdict to dict

    def to_nested_dict(self):
        return convert_to_nested_dict(self._results_nested)

    def to_pickle(self, filename, overwrite=False):
        if os.path.exists(filename):
            if not overwrite:
                logger.warn(f"Not overwriting existing file: '{filename}' (use 'overwrite=True' to overwrite)")
                return
            else:
                logger.warn(f"Overwriting existing file: '{filename}'")

        with open(filename, "wb") as f:
            pickle.dump(self, f)

    @classmethod
    def from_pickle(cls, filename):
        with open(filename, "rb") as f:
            return pickle.load(f)


def prepare_analysis_inputs(repertoire_and_genre, mode, *, cfg, min_length_monomodal_sections=3, filename_pattern=None):
    pieces = cfg.load_pieces(repertoire_and_genre, pattern=filename_pattern)
    mode = ModalCategoryType(mode)

    enforce_same_ambitus = {"final": False, "final_and_ambitus": True}[mode]

    if repertoire_and_genre == "plainchant_sequences":
        return extract_monomodal_sections(
            pieces, enforce_same_ambitus=enforce_same_ambitus, min_length=min_length_monomodal_sections
        )
    else:
        raise NotImplementedError(f"Unsupported repertoire/genre: {repertoire_and_genre}")


def accumulate_results(
    existing_results, cfg, repertoire_and_genre, analysis, unit, mode, min_length_monomodal_sections=3, p_cutoff=0.7
):
    results = existing_results or AnalysisResultCollection()
    analysis_inputs = prepare_analysis_inputs(
        repertoire_and_genre, mode, cfg=cfg, min_length_monomodal_sections=3, filename_pattern=None
    )

    grouping = GroupingByModalCategory(analysis_inputs, group_by=mode)
    logger.debug(f"Calculating results for {grouping}")
    for key in tqdm(grouping.keys):
        modal_category = grouping[key]
        df = modal_category.make_results_dataframe(analysis_func=AnalysisFuncPCFreqs("rel_freqs"), unit=unit)
        dendrogram = Dendrogram(df, p_threshold=p_cutoff)

        results.insert_results(repertoire_and_genre, analysis, unit, modal_category, "dendrogram", dendrogram.to_json())
        results.insert_results(
            repertoire_and_genre,
            analysis,
            unit,
            modal_category,
            "clusters",
            [c.to_json() for c in dendrogram.nodes_below_cutoff],
        )

    return results
