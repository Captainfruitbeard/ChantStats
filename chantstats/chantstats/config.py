from .analysis_spec import RepertoireAndGenreType


class ChantStatsConfig:
    def __init__(self, musicxml_paths):
        self.musicxml_paths = {RepertoireAndGenreType(key): path for (key, path) in musicxml_paths.items()}

    def get_musicxml_path(self, repertoire_and_genre):
        rep_and_genre = RepertoireAndGenreType(repertoire_and_genre)
        musicxml_path = self.musicxml_paths[rep_and_genre]
        return musicxml_path

    def load_pieces(self, repertoire_and_genre, *, pattern=None):
        raise DeprecationWarning("This method is deprecated. Use the function 'load_pieces()' instead.")
