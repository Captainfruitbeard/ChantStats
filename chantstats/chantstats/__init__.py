from .analysis_spec import FullAnalysisSpec
from .config import ChantStatsConfig
from .dendrogram import plot_dendrogram
from .logging import logger
from .modal_category import GroupingByModalCategory
from .plainchant_sequence_piece import load_plainchant_sequence_pieces, load_pieces
from .plainchant_sequence_monomodal_sections import MonomodalSection, extract_monomodal_sections
from .phrase_collection import PhraseCollection
from .results_export_OLD import export_dendrogram_and_stacked_bar_chart

__all__ = [
    "FullAnalysisSpec",
    "PhraseCollection",
    "GroupingByModalCategory",
    "ChantStatsConfig",
    "load_pieces",
    "load_plainchant_sequence_pieces",
    "logger",
    "chantstats_logger",
    "plot_dendrogram",
    "export_dendrogram_and_stacked_bar_chart",
]

chantstats_logger = logger  # alias for convenience

# Set __version__ attribute (via versioneer)
from ._version import get_versions

__version__ = get_versions()["version"]
del get_versions
