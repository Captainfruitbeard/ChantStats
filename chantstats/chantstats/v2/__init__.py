from .calculate_results import calculate_results
from .config import ChantStatsConfig
from .export_results import export_results
from .logging import logger
from .plainchant_sequence_piece import load_pieces
from .modal_category import GroupingByModalCategory
