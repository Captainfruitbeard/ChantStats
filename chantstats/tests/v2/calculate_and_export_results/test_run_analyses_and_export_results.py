import os
import pytest
from approvaltests.approvals import verify

from .context import chantstats
from chantstats.v2 import ChantStatsConfig, calculate_results, export_results, logger
from chantstats.v2.utils import list_directory_tree


@pytest.mark.slow
def test_run_analyses_and_export_results(tmpdir, diff_reporter):
    output_root_dir = os.getenv("CHANTSTATS_OUTPUT_ROOT_DIR", str(tmpdir))
    logger.info(f"Using output root dir: '{output_root_dir}'")

    cfg = ChantStatsConfig.from_env()
    repertoire_and_genre = "plainchant_sequences"
    min_num_phrases_per_monomodal_section = 3

    # Calculate results for PC frequencies
    results_pc_freqs = calculate_results(
        repertoire_and_genre,
        analysis="pc_freqs",
        cfg=cfg,
        min_num_phrases_per_monomodal_section=min_num_phrases_per_monomodal_section,
    )

    # Calculate results for PC tendencies
    results_pc_tendencies = calculate_results(
        repertoire_and_genre,
        analysis="pc_tendencies",
        cfg=cfg,
        min_num_phrases_per_monomodal_section=min_num_phrases_per_monomodal_section,
    )

    # Export all results
    for p_cutoff in [0.4]:
        export_results(results_pc_freqs, output_root_dir, p_cutoff=p_cutoff)
        export_results(results_pc_tendencies, output_root_dir, p_cutoff=p_cutoff)

    exported_files = list_directory_tree(output_root_dir)
    verify(exported_files, diff_reporter)
