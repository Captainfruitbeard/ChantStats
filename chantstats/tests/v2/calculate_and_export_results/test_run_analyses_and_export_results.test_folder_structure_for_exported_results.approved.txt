.
`-- p_cutoff_0.40
    `-- chant
        |-- leaps_and_melodic_outlines_L5M5
        |   `-- sequences
        |       |-- mode_degrees
        |       |   |-- dendrogram__2.authentic_2.E.png
        |       |   `-- stacked_bar_chart__2.authentic_2.E.png
        |       `-- pcs
        |           |-- dendrogram__2.authentic_2.E.png
        |           `-- stacked_bar_chart__2.authentic_2.E.png
        |-- pc_freqs
        |   `-- sequences
        |       |-- mode_degrees
        |       |   |-- dendrogram__1.final_8.C.png
        |       |   `-- stacked_bar_chart__1.final_8.C.png
        |       `-- pcs
        |           |-- dendrogram__1.final_8.C.png
        |           `-- stacked_bar_chart__1.final_8.C.png
        `-- tendency
            `-- sequences
                |-- mode_degrees
                |   |-- dendrogram__2.authentic_2.E.png
                |   |-- stacked_bar_chart__2.authentic_2.E__01.png
                |   `-- stacked_bar_chart__2.authentic_2.E__02.png
                `-- pcs
                    |-- dendrogram__2.authentic_2.E.png
                    |-- stacked_bar_chart__2.authentic_2.E__01.png
                    `-- stacked_bar_chart__2.authentic_2.E__02.png

14 directories, 14 files
